src/replication/replication_manager.cpp

```diff
@@ -0,0 +1,290 @@
+      break;
+    }
+    case MessageType::HEARTBEAT:
+      REPLICATION_LOG_TRACE(fmt::format("Heartbeat from: {}", msg.GetRoutingId()));
+      break;
+    case MessageType::REPLICATE_BUFFER: {
+      // Parse the buffer from the received message and add it to the provided replication logs.
+      {
+        nlohmann::json message = nlohmann::json::parse(msg.GetMessage());
+        // TODO(WAN): Sanity-check the received message.
+
+        uint64_t source_callback_id = msg.GetSourceCallbackId();
+        uint64_t msg_id = message.at("buf_id");
+        REPLICATION_LOG_TRACE(fmt::format("ReplicateBuffer from: {} {}", msg.GetRoutingId(), msg_id));
+
+        // Check if the message needs to be buffered.
+        if (msg_id > last_record_received_id_ + 1) {
+          // The message should be buffered if there are gaps in between the last seen buffer.
+          received_message_queue_.push(msg);
+        } else {
+          // Otherwise, pull out the log record from the message and hand them to the replication log provider.
+          // TODO(WAN): And now we parse the same JSON repeatedly.. we could parse once and store <uint64,json> instead?
+          std::string content = nlohmann::json::from_cbor(message["content"].get<std::vector<uint8_t>>());
+          provider_->AddBufferFromMessage(source_callback_id, content);
+          last_record_received_id_ = msg_id;
+          // This may unleash the rest of the buffered messages.
+          while (!received_message_queue_.empty()) {
+            auto &temporary = received_message_queue_.top();
+            source_callback_id = temporary.GetSourceCallbackId();
+            message = nlohmann::json::parse(temporary.GetMessage());
+            msg_id = message.at("buf_id");
+            // Stop once you're missing a buffer.
+            if (msg_id > last_record_received_id_ + 1) {
+              break;
+            }
+            // Otherwise, send the top buffer's contents along.
+            received_message_queue_.pop();
+            NOISEPAGE_ASSERT(msg_id == last_record_received_id_ + 1, "Duplicate buffer? Old buffer?");
+            content = nlohmann::json::from_cbor(message["content"].get<std::vector<uint8_t>>());
+            provider_->AddBufferFromMessage(source_callback_id, content);
+            last_record_received_id_ = msg_id;
+          }
+        }
+      }
+      // TODO(WAN): Is it actually necessary to buffer the message first? A crashed replica is indistinguishable from
+      //  a replica that just came up right? If buffering is not necessary, move this up so that replying is faster.
+      //  We want to reply as quickly as possible -- the sender may be waiting for our confirmation.
+      // TODO(WAN): To support async, acknowledge receipt of the buffer to the sender here.
+      // TODO(WAN): Add a size and checksum to message.
+      break;
+    }
+    default:
+      break;
+  }
+}
+
+void ReplicationManager::ReplicaHeartbeat(const std::string &replica_name) {
```

Comment:
It doesn't seem like we actually mark the replica dead as the Doxygen suggests. It seems like we just log and move on. Do we have a plan for dead replicas?

Reply:
~~The current plan is to not have dead replicas~~

We're not particularly focused on fault tolerance.
The heartbeat stuff is stuff that hasn't been tested much. A lot of assumptions are being made about replicas always being available right now. I'll see how the code shakes out after the other refactors.

Reply:
I am inclined to punt on this issue and run with the assumption that replicas won't crash without a good reason (like segfaulting etc). thoughts? in particular, please stop me if you see something problematic with the design that we can't bolt on fault tolerant stuff later.

Reply:
I think that's a fine approach, I don't think there's anything inherently wrong with the design.